### ISSUE TITLE
Rustdoc doesn't see trait implementations inside of functions

### ISSUE THREAD

**Issue by RReverser** [rust-lang/rust#41480]
It's possible to define trait implementations inside of functions in Rust. This is useful not too often, but particularly nice if you have a macro that both defines trait implementation and calls some initialization for a given type.

Anyway, the following code is valid in Rust itself:

```rust
pub struct S;

pub fn a() {
    impl From<u8> for S {
        fn from(_: u8) -> S {
            S
        }
    }
}
```

But Rustdoc doesn't generate documentation for a given trait, despite it being visible to Rust itself, and being immediately available and unique across entire crate.

I can move `impl` out of the function and then it gets seen by Rustdoc, but this is rather inconvenient when `impl` is generated as part of a bigger macro.

**Comment by GuillaumeGomez** [comment]
The scope of the implementation is only inside the function and so cannot be used outside of it and so shouldn't be documented. Also, macro != functions. If you make an implementation inside a macro, it's documented.

Closing it then, don't hesitate to reopen it if I missed something.

**Comment by RReverser** [comment]
> The scope of the implementation is only inside the function and so cannot be used outside of it and so shouldn't be documented.

No, this is exactly the problem I was referring to - `impl` is _not_ scoped. In Rust itself it's visible across entire codebase despite being declared inside of a function.

> Also, macro != functions. If you make an implementation inside a macro, it's documented.

I referred to macro only because I'm using them inside of such function to generate _both_ `impl` and call some other function.

**Comment by RReverser** [comment]
To enhance the example above:

```rust
pub struct S;

pub fn a() {
    impl From<u8> for S {
        fn from(_: u8) -> S {
            S
        }
    }
}

fn main() {
    let s = S::from(10u8); // From works here as expected, as well as in other crates using it
}
```

@GuillaumeGomez

**Comment by petrochenkov** [comment]
>don't hesitate to reopen it if I missed something

(GitHub doesn't allow non-members to reopen issues closed by others.)

**Comment by RReverser** [comment]
> doesn't allow non-members to reopen issues closed by others

Indeed, thanks for reopening this!

**Comment by GuillaumeGomez** [comment]
Thanks for reopening it @petrochenkov.

@RReverser: That's completely crazy. Another example just for the fun:

```rust
pub struct S;

trait Foo {
    fn foo(&self) {}
}

pub fn a() {
    impl Foo for S {
    }
}

fn main() {
    S.foo();
}
```

So rustdoc should document these implementations as well. Thanks for opening the issue!

**Comment by Marwes** [comment]
Got bit by this. rustdoc were simply unable to find a trait implementation defined in a function which `cargo build` picked up fine https://docs.rs/crate/serde_state/0.4.2/builds/77728 . Thought I was going crazy for second there.

**Comment by QuietMisdreavus** [comment]
I've opened https://github.com/rust-lang/rust/pull/53162 to fix this.
